minifi/minifi-nar-bundles/minifi-framework-bundle/minifi-framework/minifi-framework-core/src/main/java/org/apache/nifi/minifi/c2/command/syncresource/DefaultSyncResourceStrategy.java

```diff
@@ -0,0 +1,212 @@
+import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
+import static java.util.Map.entry;
+import static java.util.Optional.empty;
+import static java.util.Optional.ofNullable;
+import static java.util.UUID.randomUUID;
+import static java.util.function.Predicate.not;
+import static java.util.stream.Collectors.toSet;
+import static org.apache.commons.io.file.PathUtils.createParentDirectories;
+import static org.apache.nifi.c2.protocol.api.C2OperationState.OperationState.FULLY_APPLIED;
+import static org.apache.nifi.c2.protocol.api.C2OperationState.OperationState.NOT_APPLIED;
+import static org.apache.nifi.c2.protocol.api.C2OperationState.OperationState.NO_OPERATION;
+import static org.apache.nifi.c2.protocol.api.C2OperationState.OperationState.PARTIALLY_APPLIED;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import org.apache.nifi.c2.client.service.operation.SyncResourceStrategy;
+import org.apache.nifi.c2.protocol.api.C2OperationState.OperationState;
+import org.apache.nifi.c2.protocol.api.ResourceItem;
+import org.apache.nifi.c2.protocol.api.ResourcesGlobalHash;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DefaultSyncResourceStrategy implements SyncResourceStrategy {
+
+    private static final Logger LOG = LoggerFactory.getLogger(DefaultSyncResourceStrategy.class);
+
+    private static final Set<Entry<OperationState, OperationState>> SUCCESS_RESULT_PAIRS = Set.of(
+        entry(NO_OPERATION, NO_OPERATION),
+        entry(NO_OPERATION, FULLY_APPLIED),
+        entry(FULLY_APPLIED, NO_OPERATION),
+        entry(FULLY_APPLIED, FULLY_APPLIED));
+
+    private static final Set<Entry<OperationState, OperationState>> FAILED_RESULT_PAIRS = Set.of(
+        entry(NO_OPERATION, NOT_APPLIED),
+        entry(NOT_APPLIED, NO_OPERATION),
+        entry(NOT_APPLIED, NOT_APPLIED));
+
+    private final ResourceRepository resourceRepository;
+    private final Path assetDirectory;
+    private final Path extensionDirectory;
+
+    public DefaultSyncResourceStrategy(ResourceRepository resourceRepository, Path assetDirectory, Path extensionDirectory) {
+        this.resourceRepository = resourceRepository;
+        this.assetDirectory = assetDirectory;
+        this.extensionDirectory = extensionDirectory;
+    }
+
+    @Override
+    public OperationState synchronizeResourceRepository(ResourcesGlobalHash c2GlobalHash, List<ResourceItem> c2ServerItems,
+                                                        BiFunction<String, Function<InputStream, Optional<Path>>, Optional<Path>> resourceDownloadFunction,
+                                                        Function<String, Optional<String>> urlEnrichFunction) {
+        Set<ResourceItem> c2Items = c2ServerItems.stream().collect(toSet());
```

Comment:
There 2 collects methods could be replaced with new HashSet<>()

Reply:
Refactored